change grep to respect file order

elee1766 opened the pull request:
this pr does 2 things

first it refactors cmd to split the logic into multiple parts. this made the logic a bit easier for me to follow. 

second, instead of using nGrep to keep track of the amount of greps remaining to process, use the allGrep channel, and just close it when work is done. this should preserve the file order.

im not really sure the best way to create a test to simulate the race... the existing tests pass and from usage before and after, it seems that file order is being respected which is good. let me know if you want me to try to make something.

rminnich commented:
so, hey, that whole mess with that chans and all was me messing around in 2012 to see if it worked. I am not sure it was such a great approach, would appreciate your thoughts. Do you think that approach is still even worth it?

elee1766 commented:
> so, hey, that whole mess with that chans and all was me messing around in 2012 to see if it worked. I am not sure it was such a great approach, would appreciate your thoughts. Do you think that approach is still even worth it?

I don't think the channel is doing much here. since files are both walked through and opened in a single thread, we aren't actually taking advantage of the buffered "treeNames" channel properly. you can prove this yourself by running grep on a large directory - you'll notice that it never goes above one cpu. there's a tiny possible bit of io latency saved by using the channels for time to first print, not really sure how to time that but i don't think its significant.

so honestly, i would say should either

1. keep it single threaded, get rid of channels, and just do the file open and grep in the os.Walk. it would make the code a lot simpler.
2. make it properly multithreaded, which would probably greatly increase the speed of grep in more powerful computers (at the cost of more memory use)

i dont really know u-root project well, but if the goal is to keep tools simple & low resource, probably I would go with the first option. it would be pretty trivial and probably reduce the line count by 50-100

MDr164 commented:
> i dont really know u-root project well, but if the goal is to keep tools simple & low resource, probably I would go with the first option. it would be pretty trivial and probably reduce the line count by 50-100

The primary use-case for u-root is to be used as an initramfs for LinuxBoot or similar projects. It could as well be used in other contexts as well but that's up to the users then. So given that u-roots main use-case is in a size contraint and critical path of the machines boot up I would argue that keeping it simple and low resource intense is the way to go. What do the others think?

rafaelcn commented on cmds/core/grep/grep.go:
```diff
@@ -204,68 +199,85 @@ func (c *cmd) run() error {
-			c.nGrep--
-			if c.nGrep == 0 {
-				break
+func (c *cmd) run_print() error {
```
I don't think snake case is the right way of writing idiomatic Go code.  Please take a look at the other functions as well.

By the way, I love snake case.

elee1766 replied:
ahhh yeah in my org we use snake case for single use unexported functions and lowerCamelCase for multiple use. 

i can change everything (but we should probably reach a decision on the concurrency thing)

rafaelcn commented:
> [...] So given that u-roots main use-case is in a size contraint and critical path of the machines boot up I would argue that keeping it simple and low resource intense is the way to go. What do the others think?

I agree that we should keep the resource usage low

elee1766 commented:
happy to change this PR to simplify the channel and move the grep into the os.Walk. 

let me know.

rminnich commented:
thanks for looking at this code. Yeah, go ahead and simplify it. Also, please, record binary size before and after for posterity :-)

elee1766 commented:
okay. i've refactored the entire thing to be a lot simpler. I also found that the implementation in `main` races or something, it's missing lines when compared to  my system grep and the current version 

1. back to 1 function, since the code is so short
2. change the use of fmt.Sprintf -> writing to a bufio.Writer. the bufio.Writer will flush after every line.
3. changed bufio.Reader -> bufio.Scanner. This allows 0 allocation reads from the file
3. change use of strings -> bytes, so things stay 0 alloc

binary size is smaller after:
```
go version go1.20.2 linux/amd64
main branch: 2,649,623
grep-order branch:  2,646,130
diff - 3493 
```

for a simple performance test, I tested by grepping the drivers folder of the linux kernel. (i warmed up the dataset multiple times in page cache first. my system has enough memory to fit the entire directory in memory)

for this dataset on my computer, it was 10 times faster than `main` branch, now only 2-3 times slower than gnu grep

i investigated and found that the `main` grep missed a line that it should have printed, somehow


system grep:
```
980247
grep -r "int" ~/repo/github.com/torvalds/linux/drivers  0.41s user 0.18s system 99% cpu 0.595 total
wc -l  0.02s user 0.02s system 7% cpu 0.595 total
```

`main` branch:
```
time ./grep -r "int" ~/repo/github.com/torvalds/linux/drivers | wc -l                 
980246
./grep -r "int" ~/repo/github.com/torvalds/linux/drivers  11.66s user 1.06s system 314% cpu 4.039 total
wc -l  0.08s user 0.56s system 15% cpu 4.039 total
```

`grep-order` branch:
```
time ./grep -r "int" ~/repo/github.com/torvalds/linux/drivers | wc -l 
 980247
./grep -r "int" ~/repo/github.com/torvalds/linux/drivers  0.95s user 0.49s system 101% cpu 1.418 total
wc -l  0.07s user 0.36s system 30% cpu 1.418 total

```

rminnich commented:
very nice! but oh no! Conflicts! can you take one more pass? This is a very nice improvement.

elee1766 commented:
> very nice! but oh no! Conflicts! can you take one more pass? This is a very nice improvement.

weird git thing i guess. i think it should be okay now?

rminnich commented:
nope. I think you'll need to do a rebase and make sure it works. Prob a good idea to squash some commits anyway, I doubt we need all of them?

elee1766 commented:
hmm i just did a rebase and its still angry. ill open a new pr in a new branch.
